lsif/src/xrepo.ts

```diff
@@ -24,6 +24,49 @@ const insertionMetrics = {
     errorsCounter: xrepoQueryErrorsCounter,
 }
 
+/**
+ * A CTE definition that gets all ancestors of a commit recursively.
+ */
+const lineage_ancestors = `
+    -- Get all ancestors of the tip
+    WITH RECURSIVE lineage(id, repository, "commit", parent) AS (
+        SELECT c.* FROM lsif_commits c WHERE c.repository = $1 AND c."commit" = $2
+        UNION
+        SELECT c.* FROM lineage a JOIN lsif_commits c ON a.repository = c.repository AND a.parent = c."commit"
+    ),
+`
+
+/**
+ * A set of CTE definitions that prune LSIF dump identifiers so that only the ones that
+ * are visible from a source commit remain. This assumes that a CTE definition `lineage`
+ * is available earlier in the query. The CTE `lineage_fanout` (defined above) can be used
+ * as such a CTE definition.
+ */
+const ctes = `
+    -- Limit the visibility to the maximum traversal depth and approximate
+    -- each commit's depth by its row number.
+    limited_lineage AS (
+        SELECT a.*, row_number() OVER() as n from lineage a LIMIT $3
```

Comment:
```suggestion
        SELECT a.*, row_number() OVER() as n from lineage a LIMIT $traversalLimit
```

After extracting these into the global scope, it's easier to mix up the parameter order. Do you know if typeorm supports named parameters in `query()`? Based on the types, it looks like it doesn't

![image](https://user-images.githubusercontent.com/1387653/68152862-b64c1380-ff01-11e9-9c7b-32c884d4860a.png)

Maybe there's another query interface that supports named parameters?

Alternatively, how about turning this into a function that returns a raw SQL string?

Reply:
I'd like to transition to sequelize at some point, I think. It's more popular and has a better interface.

I can functionize it.